File: go/0059M-螺旋矩阵II.go
package main

// 左闭右开
func generateMatrix(n int) [][]int {
	matrix := make([][]int, n)
	for i := 0; i < n; i++ {
        matrix[i] = make([]int, n)
    }
	startx, starty := 0, 0 // 维护每次循环的起始位置
	loop := n / 2          // loop维护循环次数
	cnt, offset := 1, 1    // cnt维护位置数值, offset维护遍历长度
	i, j := 0, 0
	for loop != 0 {
		loop--
		i = startx
		j = starty

		for j = starty; j < starty+n-offset; j++ {
			matrix[i][j] = cnt
			cnt++
		}

		for i = startx; i < startx+n-offset; i++ {
			matrix[i][j] = cnt
			cnt++
		}

		for ; j > starty; j-- {
			matrix[i][j] = cnt
			cnt++
		}

		for ; i > startx; i-- {
			matrix[i][j] = cnt
			cnt++
		}

		// 循环起始位置主对角线方向移动
		startx++
		starty++
		offset += 2
	}
	if n&1 == 1 { // 奇数中间未遍历
		mid := n / 2
		matrix[mid][mid] = cnt
	}
	return matrix
}

File: cpp/0059M-螺旋矩阵II.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-13 09:23:21
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

using namespace std;

// 左闭右开
// 按圈循环
class Solution {
   public:
    vector<vector<int>> generateMatrix(int n) {
        int startx, starty, loop, cnt, offset, i, j;
        vector<vector<int>> matrix(n, vector<int>(n, 0));
        startx = starty = i = j = 0;
        cnt = offset = 1;
        loop = n / 2;
        while (loop--) {
            i = startx;
            j = starty;

            for (; j < starty + n - offset; j++) matrix[i][j] = cnt++;  // 👉→
            for (; i < startx + n - offset; i++) matrix[i][j] = cnt++;  // 👇↓
            for (; j > starty; j--) matrix[i][j] = cnt++;  // 👈←
            for (; i > startx; i--) matrix[i][j] = cnt++;  // 👆↑

            startx++, starty++;  // 向内一圈
            offset += 2;         // 两个外圈长度
        }
        if (n & 1)  // 奇数中间未遍历
            matrix[n / 2][n / 2] = cnt;
        return matrix;
    }
};
